File: app/src/main/java/com/originb/inkwisenote2/modules/fileexplorer/DirectoryExplorerViewModel.java
package com.originb.inkwisenote2.modules.fileexplorer;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.originb.inkwisenote2.modules.backgroundjobs.BackgroundOps;

import java.io.File;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class DirectoryExplorerViewModel extends ViewModel {

    private final MutableLiveData<List<FileGroup>> _fileGroups = new MutableLiveData<>();
    public final LiveData<List<FileGroup>> fileGroups = _fileGroups;

    private final MutableLiveData<Boolean> _isRefreshing = new MutableLiveData<>();
    public final LiveData<Boolean> isRefreshing = _isRefreshing;

    private final MutableLiveData<String> _toolbarTitle = new MutableLiveData<>();
    public final LiveData<String> toolbarTitle = _toolbarTitle;

    private final MutableLiveData<String> _toastMessage = new MutableLiveData<>();
    public final LiveData<String> toastMessage = _toastMessage;

    private File currentDirectory;
    private final Stack<File> navigationHistory = new Stack<>();

    // Timestamp pattern regex
    private static final Pattern TIMESTAMP_PATTERN = Pattern.compile("^(\\d{8,14})(?:[\\.-]|$)");

    /**
     * Initialize the explorer with a starting directory.
     */
    public void init(File initialDir) {
        if (currentDirectory == null) {
            currentDirectory = initialDir;
            loadCurrentDirectory();
        }
    }

    public void loadCurrentDirectory() {
        if (currentDirectory == null) return;

        _isRefreshing.setValue(true);
        _toolbarTitle.setValue(currentDirectory.getName().isEmpty() ? "Files" : currentDirectory.getName());

        BackgroundOps.execute(() -> {
            List<FileItem> fileItems = new ArrayList<>();
            File[] files = currentDirectory.listFiles();

            if (files != null && files.length > 0) {
                for (File file : files) {
                    fileItems.add(new FileItem(file));
                }
                return groupFilesByTimestamp(fileItems);
            }
            return new ArrayList<FileGroup>();
        }, result -> {
            _fileGroups.setValue(result);
            _isRefreshing.setValue(false);
        });
    }

    /**
     * Grouping logic moved from Activity to ViewModel for testability.
     */
    private List<FileGroup> groupFilesByTimestamp(List<FileItem> fileItems) {
        Map<String, List<FileItem>> groupMap = new HashMap<>();
        List<FileItem> ungroupedFiles = new ArrayList<>();

        for (FileItem item : fileItems) {
            if (item.isDirectory()) {
                ungroupedFiles.add(item);
                continue;
            }

            String timestamp = extractTimestamp(item.getName());
            if (timestamp.isEmpty()) {
                ungroupedFiles.add(item);
            } else {
                if (!groupMap.containsKey(timestamp)) {
                    groupMap.put(timestamp, new ArrayList<>());
                }
                groupMap.get(timestamp).add(item);
            }
        }

        List<FileGroup> result = new ArrayList<>();
        for (Map.Entry<String, List<FileItem>> entry : groupMap.entrySet()) {
            result.add(new FileGroup(entry.getKey(), entry.getValue()));
        }
        for (FileItem item : ungroupedFiles) {
            result.add(new FileGroup(item));
        }

        Collections.sort(result, (a, b) -> {
            if (a.isDirectory() && !b.isDirectory()) return -1;
            if (!a.isDirectory() && b.isDirectory()) return 1;
            if (a.isGroup() && !b.isGroup()) return -1;
            if (!a.isGroup() && b.isGroup()) return 1;
            if (a.isGroup() && b.isGroup()) return b.getTimestamp().compareTo(a.getTimestamp());
            return a.getGroupName().compareToIgnoreCase(b.getGroupName());
        });

        return result;
    }

    private String extractTimestamp(String filename) {
        if (filename == null) return "";
        Matcher matcher = TIMESTAMP_PATTERN.matcher(filename);
        return matcher.find() ? matcher.group(1) : "";
    }

    /**
     * Navigation Logic
     */
    public void navigateInto(File directory) {
        navigationHistory.push(currentDirectory);
        currentDirectory = directory;
        loadCurrentDirectory();
    }

    public boolean navigateBack() {
        if (!navigationHistory.isEmpty()) {
            currentDirectory = navigationHistory.pop();
            loadCurrentDirectory();
            return true;
        }
        return false;
    }

    /**
     * Delete Logic
     */
    public void deleteFileGroup(FileGroup fileGroup) {
        BackgroundOps.execute(() -> {
            boolean success = true;
            if (fileGroup.isGroup()) {
                for (File file : fileGroup.getAllRawFiles()) {
                    if (!file.delete()) success = false;
                }
            } else {
                FileItem singleFile = fileGroup.getPrimaryFile();
                if (singleFile != null) {
                    if (singleFile.isDirectory()) {
                        success = deleteRecursive(singleFile.getFile());
                    } else {
                        success = singleFile.getFile().delete();
                    }
                }
            }
            return success;
        }, success -> {
            if (success) {
                _toastMessage.setValue("Deleted successfully");
                loadCurrentDirectory();
            } else {
                _toastMessage.setValue("Failed to delete some items");
            }
        });
    }

    private boolean deleteRecursive(File fileOrDirectory) {
        if (fileOrDirectory.isDirectory()) {
            File[] children = fileOrDirectory.listFiles();
            if (children != null) {
                for (File child : children) {
                    if (!deleteRecursive(child)) return false;
                }
            }
        }
        return fileOrDirectory.delete();
    }

    public File getCurrentDirectory() {
        return currentDirectory;
    }
}

File: app/src/main/java/com/originb/inkwisenote2/modules/fileexplorer/DirectoryExplorerViewModel.kt
package com.originb.inkwisenote2.modules.fileexplorer

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.originb.inkwisenote2.modules.backgroundjobs.BackgroundOps
import java.io.File
import java.util.*
import java.util.concurrent.Callable
import java.util.function.Consumer
import java.util.regex.Matcher
import java.util.regex.Pattern

class DirectoryExplorerViewModel : ViewModel() {
    private val _fileGroups = MutableLiveData<MutableList<FileGroup?>?>()
    val fileGroups: LiveData<MutableList<FileGroup?>?> = _fileGroups

    private val _isRefreshing = MutableLiveData<Boolean?>()
    val isRefreshing: LiveData<Boolean?> = _isRefreshing

    private val _toolbarTitle = MutableLiveData<String?>()
    val toolbarTitle: LiveData<String?> = _toolbarTitle

    private val _toastMessage = MutableLiveData<String?>()
    val toastMessage: LiveData<String?> = _toastMessage

    var currentDirectory: File? = null
        private set
    private val navigationHistory = Stack<File?>()

    /**
     * Initialize the explorer with a starting directory.
     */
    fun init(initialDir: File?) {
        if (currentDirectory == null) {
            currentDirectory = initialDir
            loadCurrentDirectory()
        }
    }

    fun loadCurrentDirectory() {
        if (currentDirectory == null) return

        _isRefreshing.setValue(true)
        _toolbarTitle.setValue(if (currentDirectory!!.name.isEmpty()) "Files" else currentDirectory!!.name)

        BackgroundOps.execute(Callable {
            val fileItems = ArrayList<FileItem>()
            val files = currentDirectory!!.listFiles()
            if (files != null && files.isNotEmpty()) {
                for (file in files) {
                    fileItems.add(FileItem(file))
                }
                groupFilesByTimestamp(fileItems)
            } else {
                ArrayList<FileGroup?>()
            }
        }, Consumer { result: MutableList<FileGroup?>? ->
            _fileGroups.setValue(result)
            _isRefreshing.setValue(false)
        })
    }

    /**
     * Grouping logic moved from Activity to ViewModel for testability.
     */
    private fun groupFilesByTimestamp(fileItems: MutableList<FileItem>): MutableList<FileGroup?> {
        val groupMap = HashMap<String?, MutableList<FileItem>>()
        val ungroupedFiles = ArrayList<FileItem>()

        for (item in fileItems) {
            if (item.isDirectory) {
                ungroupedFiles.add(item)
                continue
            }
            val timestamp = extractTimestamp(item.name)
            if (timestamp.isEmpty()) {
                ungroupedFiles.add(item)
            } else {
                groupMap.getOrPut(timestamp) { ArrayList() }.add(item)
            }
        }

        val result = ArrayList<FileGroup?>()
        for (entry in groupMap.entries) {
            result.add(FileGroup(entry.key, entry.value))
        }
        for (item in ungroupedFiles) {
            result.add(FileGroup(item))
        }

        Collections.sort(result, Comparator { a: FileGroup?, b: FileGroup? ->
            when {
                a!!.isDirectory && !b!!.isDirectory -> -1
                !a.isDirectory && b!!.isDirectory -> 1
                a.isGroup && !b!!.isGroup -> -1
                !a.isGroup && b!!.isGroup -> 1
                a.isGroup && b!!.isGroup -> (b.timestamp ?: "").compareTo(a.timestamp ?: "")
                else -> (a.groupName ?: "").compareTo(b!!.groupName ?: "", ignoreCase = true)
            }
        })
        return result
    }

    private fun extractTimestamp(filename: String?): String {
        if (filename == null) return ""
        val matcher: Matcher = TIMESTAMP_PATTERN.matcher(filename)
        return if (matcher.find()) matcher.group(1) else ""
    }

    /**
     * Navigation Logic
     */
    fun navigateInto(directory: File?) {
        navigationHistory.push(currentDirectory)
        currentDirectory = directory
        loadCurrentDirectory()
    }

    fun navigateBack(): Boolean {
        if (!navigationHistory.isEmpty()) {
            currentDirectory = navigationHistory.pop()
            loadCurrentDirectory()
            return true
        }
        return false
    }

    /**
     * Delete Logic
     */
    fun deleteFileGroup(fileGroup: FileGroup) {
        BackgroundOps.execute(Callable {
            var success = true
            if (fileGroup.isGroup) {
                for (file in fileGroup.allRawFiles) {
                    if (file != null && !file.delete()) success = false
                }
            } else {
                fileGroup.primaryFile?.let { singleFile ->
                    success = if (singleFile.isDirectory) {
                        deleteRecursive(singleFile.file)
                    } else {
                        singleFile.file.delete()
                    }
                }
            }
            success
        }, Consumer { success: Boolean? ->
            if (success == true) {
                _toastMessage.setValue("Deleted successfully")
                loadCurrentDirectory()
            } else {
                _toastMessage.setValue("Failed to delete some items")
            }
        })
    }

    private fun deleteRecursive(fileOrDirectory: File): Boolean {
        if (fileOrDirectory.isDirectory) {
            val children = fileOrDirectory.listFiles()
            if (children != null) {
                for (child in children) {
                    if (!deleteRecursive(child)) return false
                }
            }
        }
        return fileOrDirectory.delete()
    }

    companion object {
        // Timestamp pattern regex
        private val TIMESTAMP_PATTERN: Pattern = Pattern.compile("^(\\d{8,14})(?:[\\.-]|$)")
    }
}